### PR TITLE
PLANNER-2399 Improve performance of CS-D groupBy

### PR DESCRIPTION
**Pull request by triceo**
Brings improvements of a couple of percentage points.
Depends on https://github.com/kiegroup/optaplanner/pull/1256.

### CODE REVIEW

File: build/optaplanner-build-parent/pom.xml
Comment (triceo) [inline build/optaplanner-build-parent/pom.xml:26]: This will go aways before merging. (When the bump PR is merged.)

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/common/AbstractRuleContext.java
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/common/AbstractRuleContext.java:40]: The explanation was no longer necessary as the context was hidden from people and slightly repurposed, so I removed it.

**Review by rsynek — APPROVED** [review]
LGTM, but let's wait for the CI and SonarCloud.

**Pull request merged**
